### PR TITLE
Add Spring interpolation

### PR DESCRIPTION
**Pull request by mrousavy**
This PR adds a custom interpolation mode: `'spring'` for element transitions, shared element transitions and animations in general, ref https://github.com/wix/react-native-navigation/issues/6431

### iOS

Implemented a custom damped harmonic spring interpolation for iOS using [this grapher file](https://github.com/jenox/UIKit-Playground/tree/master/01-Demystifying-UIKit-Spring-Animations/Resources/DampedHarmonicSpring.gcx) for reference, which uses a custom `springiness` parameter (ζ) and `mass` parameter (m).

### Android

I used [OvershootInterpolator](https://developer.android.com/reference/android/view/animation/OvershootInterpolator) from the Android animation API, which only has a `tension` property to configure `springiness`.

### Demo

<table>
<tr>
<th>Android</th>
<th>iOS</th>
</tr>
<tr>
<td width="50%"
></td>
<td width="50%"></td>
</tr>
</table>

### TODO

1. Due to the differences in the two APIs, we have to find a matching configuration and parse/map that on each platform, so it looks similarily. I'm currently hard-coding the parameters (iOS: `springiness: 0.3`, `mass: 3`, Android: `tension: 1.0`), but they have to be exposed to the JS-API as-well. I'll leave that part up to you at wix, since that changes the public API. (See #6431 for my suggestions regarding the public TS API)
2. On iOS, I'd suggest instead of using single static functions for interpolation, we use an object-oriented approach to create interpolation instances. Then, just use the `interpolationInstance.interpolate(p)` function on that object to actually interpolate. This makes the API more extensible, and also allows for us to "cache" the parameters I'm creating on every call to `RNNSpring()` ([those lines right here](https://github.com/mrousavy/react-native-navigation/blob/3e0f1ea0cea4fc04072ee7cc8ac6ddc793df7a1c/lib/ios/RNNInterpolator.m#L42-L53))

### CODE REVIEW

**Comment by mrousavy** [comment]
Also, why does [this property](https://github.com/wix/react-native-navigation/blob/master/lib/src/interfaces/Options.ts#L900) not have other interpolations such as `'linear'`, `'accelerateDecelerate'`, etc.?

**Review by yogevbd — APPROVED** [review]

**Comment by guyca** [comment]
> Also, why does [this property](https://github.com/wix/react-native-navigation/blob/master/lib/src/interfaces/Options.ts#L900) not have other interpolations such as `'linear'`, `'accelerateDecelerate'`, etc.?

The TS interface is missing a few interpolation options, nice catch!

File: lib/ios/RNNInterpolator.m [hunk 1]
```diff
@@ -28,13 +28,39 @@ static CGFloat RNNApplyInterpolation(CGFloat p, RNNInterpolationOptions interpol
...
+            return RNNSpring(p, 0.3, 3);
     }
 }
 
 static CGFloat RNNInterpolate(CGFloat from, CGFloat to, CGFloat p, RNNInterpolationOptions interpolation) {
     return from + RNNApplyInterpolation(p, interpolation) * (to - from);
 }
 
+static CGFloat RNNSpring(CGFloat p, CGFloat springiness, CGFloat mass) {
+	// TODO: Cache those allocations
```
Comment (guyca) [inline lib/ios/RNNInterpolator.m:42]: Why do we need to cache these variables?

**Review by guyca — CHANGES_REQUESTED** [review]

File: lib/ios/RNNInterpolator.m (hunk 1 above)
Comment (mrousavy) [reply lib/ios/RNNInterpolator.m:42]: Those functions are being executed many times in a short amount of time, that's why my gut was telling me to remove everything that doesn't _have_ to be in there to somewhere else, especially "statics" like those variables to improve performance. They are being created with the values that the user supplies (calculated from `mass` and `springiness`) and won't change from there on forward.

Comment (mrousavy) [reply lib/ios/RNNInterpolator.m:42]: tbh it's just a bit of easy math, some `sqrt`s and `pow`s, nothing fancy like `cos`ines so we're probably fine and maybe the compiler will do some tricks anyways.

**Review by guyca — APPROVED** [review]

**Pull request merged**